tuning/sft_trainer.py

```diff
@@ -0,0 +1,149 @@
+        })
+    elif isinstance(tokenizer, GPTNeoXTokenizerFast):
+        tokenizer.add_special_tokens({
+            "pad_token": "<pad>",
+        })
+    
+    model_max_length = tokenizer.model_max_length
+    logger.info(f"Model max length {model_max_length}")
+    
+    # TODO: we need to change this, perhaps follow what open instruct does?
+    special_tokens_dict = dict()
+    if tokenizer.pad_token is None:
+        logger.warning("PAD token set to default, missing in tokenizer")
+        special_tokens_dict["pad_token"] = configs.DEFAULT_PAD_TOKEN
+    if tokenizer.eos_token is None:
+        logger.warning("EOS token set to default, missing in tokenizer")
+        special_tokens_dict["eos_token"] = configs.DEFAULT_EOS_TOKEN
+    if tokenizer.bos_token is None:
+        logger.warning("BOS token set to default, missing in tokenizer")
+        special_tokens_dict["bos_token"] = configs.DEFAULT_BOS_TOKEN
+    if tokenizer.unk_token is None:
+        logger.warning("UNK token set to default, missing in tokenizer")
+        special_tokens_dict["unk_token"] = configs.DEFAULT_UNK_TOKEN
+
+    tokenizer_data_utils.tokenizer_and_embedding_resize(
+        special_tokens_dict=special_tokens_dict,
+        tokenizer=tokenizer,
+        model=model,
+    )
+    
+    # load the data by parsing JSON
+    json_dataset = datasets.load_dataset('json', data_files=data_args.data_path)
+    logger.info(f"Dataset length is {len(json_dataset['train'])}")
+
+    aim_callback = get_aimstack_callback()
+    callbacks=[aim_callback,PeftSavingCallback()]
+
+    if training_args.packing:
+        logger.info("Packing is set to True")
+        trainer = SFTTrainer(
+            model=model,
+            tokenizer=tokenizer,
+            train_dataset=json_dataset['train'],
+            dataset_text_field=data_args.dataset_text_field,
+            packing=True,
+            args=training_args,
+            max_seq_length=model_max_length,
+            callbacks=callbacks,
+            peft_config=peft_config,
+        )
+    else:
+        logger.info("Packing is set to False")
+        if data_args.response_template is None:
+            logger.error("Error, response template is None, needs to be set for training")
+            exit(-1)
+        
+        if data_args.dataset_text_field is None:
+            logger.error("Error, dataset_text_field is None, needs to be set for training")
+            exit(-1)
+        
+        #the below is specific to Llama since it uses sentencepiece tokenizer
+        # TODO: the below code is preparing data in a way that was specificed in  the example code from SFTTrainer for Llama.
+        # This is not accurate for GPTBigCode.
+        response_template_ids = tokenizer.encode(data_args.response_template, add_special_tokens=False)[2:]
+        data_collator = DataCollatorForCompletionOnlyLM(response_template_ids, tokenizer=tokenizer, ignore_index=configs.IGNORE_INDEX)
+        trainer = SFTTrainer(
```

Comment:
IMO it might be simpler to `if packing: collator = None; else collator = ...` and then just call the SFTTrainer constructor once.